lib/parse_identifier.rb

```diff
@@ -0,0 +1,176 @@
+require 'identifiers'
+
+# https://github.com/altmetric/identifiers
+#
+# Identifiers::DOI.extract('example: 10.123/abcd.efghi')
+# # => ["10.123/abcd.efghi"]
+#
+# It can support many identifiers:
+# Identifiers::AdsBibcode.extract('')
+# Identifiers::ArxivId.extract('')
+# Identifiers::DOI.extract('')
+# Identifiers::Handle.extract('')
+# Identifiers::ISBN.extract('')
+# Identifiers::NationalClinicalTrialId.extract('')
+# Identifiers::ORCID.extract('')
+# Identifiers::PubmedId.extract('')
+# Identifiers::RepecId.extract('')
+# Identifiers::URN.extract('')
+
+# Parse the identifiers in a PublicationIdentifier
+#
+# This class should try to work with the input parameter as a read-only object.
+# It should only modify it when explicitly asked to `update` it.  If it's
+# an active record object, assign new values to it, but don't save it, leave
+# that persistence action to the consumer.
+class ParseIdentifier
+
+  URI_PREFIX_DOI = 'http://dx.doi.org/'.freeze
+  # URI_PREFIX_ISBN = ''.freeze # maybe there is no such thing?
+  URI_PREFIX_PMID = 'https://www.ncbi.nlm.nih.gov/pubmed/'.freeze
+
+  attr_reader :pub_id
+
+  # The pub_id parameter should respond to these fields:
+  # pub_id[:identifier_type]
+  # pub_id[:identifier_value]
+  # pub_id[:identifier_uri]
+  # @param pub_id [PublicationIdentifier|Hash]
+  def initialize(pub_id)
+    @pub_id = pub_id
```

Comment:
When you look at the runner script, this lib really only needs to expose the `#update` method and it will take care of the rest.  In that context, it doesn't need to know anything (and should not need to know anything) about what type of PublicationIdentifier it is.

Reply:
OK, will take out the `Hash` input param.

Reply:
The OOP class inheritance is an option, for sure, although a bit more work to refactor stuff rather than sniffing.  (Of course, this is still WIP and I do refactor as I go, usually.  Even so, an early review is useful to steer it along.)